### PR TITLE
Address prototype comments

### PR DESCRIPTION
**Pull request by 333fred**
Addresses remaining prototype comments and behavior updates from the LDM. Spec change here: https://github.com/dotnet/csharplang/pull/6140.

Spec: https://github.com/dotnet/csharplang/blob/main/proposals/required-members.md
Test plan: https://github.com/dotnet/roslyn/issues/57046.

### CODE REVIEW

**Comment by 333fred** [comment]
@jcouv @RikkiGibson for review.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs [hunk 1]
```diff
@@ -1634,5 +1634,41 @@ internal bool IsSimpleProgram
...
+            // We don't warn for required members from base types, as the user either has a warning that they're depending on an obsolete constructor/type already,
+            // or the original author ignored this warning.
+
+            // Obsolete states should have already been calculated by this point in the pipeline.
+            Debug.Assert(ObsoleteKind != ObsoleteAttributeKind.Uninitialized);
+            Debug.Assert(GetMembers().All(m => m.ObsoleteKind != ObsoleteAttributeKind.Uninitialized));
+
+            if (ObsoleteKind != ObsoleteAttributeKind.None
+                || GetMembers().All(m => m is not MethodSymbol { MethodKind: MethodKind.Constructor, ObsoleteKind: ObsoleteAttributeKind.None }
+                                         || !((MethodSymbol)m).ShouldCheckRequiredMembers()))
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs:1655]: might be a little easier to read if the pattern on the previous line introduced a 'method' local instead of casting here.

File: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs [hunk 2]
```diff
@@ -3688,6 +3801,9 @@ public Derived() : base()
 
         var comp = CreateCompilationWithRequiredMembers(code);
         comp.VerifyDiagnostics(
+            // (9,15): warning CS8618: Non-nullable property 'Prop1' must contain a non-null value when exiting constructor. Consider declaring the property as nullable.
+            //     protected Base() {}
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:3805]: consider updating the // 1 comments.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateMethodSymbol.cs [hunk 3]
```diff
@@ -450,5 +452,6 @@ private static bool IsUnique(ArrayBuilder<ParameterSymbol> currentParameters, st
 
             return true;
         }
+
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateMethodSymbol.cs:455]:
```suggestion
```

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 4]
```diff
@@ -7136,4 +7136,10 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="ERR_UnsupportedCompilerFeature" xml:space="preserve">
     <value>'{0}' requires compiler feature '{1}', which is not supported by this version of the C# compiler.</value>
   </data>
+  <data name="WRN_ObsoleteMembersShouldNotBeRequired" xml:space="preserve">
+    <value>Members attributed with 'ObsoleteAttribute' should not be required unless the containing type is obsolete or all constructors are obsolete.</value>
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/CSharpResources.resx:7140]: Consider including the symbol for the member which had both required and obsolete on it here, i.e. as argument `{0}`.

**Review by RikkiGibson — APPROVED** [review]

**Comment by 333fred** [comment]
@jcouv for the second review.

**Comment by 333fred** [comment]
@jcouv for the second review.

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 5]
```diff
@@ -7136,6 +7136,12 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="ERR_UnsupportedCompilerFeature" xml:space="preserve">
     <value>'{0}' requires compiler feature '{1}', which is not supported by this version of the C# compiler.</value>
   </data>
+  <data name="WRN_ObsoleteMembersShouldNotBeRequired" xml:space="preserve">
+    <value>Required member '{0}' should not be attributed with 'ObsoleteAttribute' unless the containing type is obsolete or all constructors are obsolete.</value>
+  </data>
+  <data name="WRN_ObsoleteMembersShouldNotBeRequired_Title" xml:space="preserve">
+    <value>Members attributed with 'ObsoleteAttribute' should not be required unless the containing type is obsolete or all constructors are obsolete.</value>
+  </data>
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/CSharpResources.resx:7144]: The spec update linked from OP says: "`required` is not allowed to be applied to indexers.". Will that be handled in a different PR?

File: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs [hunk 6]
```diff
@@ -3678,24 +3791,27 @@ public class Derived : Base
             
                 public Derived() : base()
                 {
-                    Prop1.ToString();
+                    Prop1.ToString(); // 1
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:3794]: The numbered comment is in the wrong location.

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 5 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/CSharpResources.resx:7144]: This is already handled.

**Comment by jcouv** [comment]
>     public void InvalidModifierCombinations()

Can we update `InvalidModifierCombinations` with other invalid combination of modifiers (for example `ref readonly`) listed in the spec update?

---
Refers to: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:216 in 6188948. [](commit_id = 618894873d5e4974f1a5ae624a3f5d25f9e4c132, deletion_comment = False)

File: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs [hunk 7]
```diff
@@ -1253,52 +1253,165 @@ class C
...
+
+        comp.VerifyDiagnostics(
+            // (14,25): warning CS9513: Required member 'C.Field' should not be attributed with 'ObsoleteAttribute' unless the containing type is obsolete or all constructors are obsolete.
+            //     public required int Field;
+            Diagnostic(ErrorCode.WRN_ObsoleteMembersShouldNotBeRequired, "Field").WithArguments("C.Field").WithLocation(14, 25),
+            // (16,25): warning CS9513: Required member 'C.Prop1' should not be attributed with 'ObsoleteAttribute' unless the containing type is obsolete or all constructors are obsolete.
+            //     public required int Prop1 { get; set; }
+            Diagnostic(ErrorCode.WRN_ObsoleteMembersShouldNotBeRequired, "Prop1").WithArguments("C.Prop1").WithLocation(16, 25)
+        );
+    }
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:1413]: The spec update calls out for property without setter/initer or with less accessible setter/initer. Do we have tests covering those?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 8]
```diff
@@ -632,11 +633,22 @@ internal override void ForceComplete(SourceLocation? locationOpt, CancellationTo
 
                             EnsureFieldDefinitionsNoted();
 
-                            // We've completed all members, so we're ready for the PointedAtManagedTypeChecks;
-                            // proceed to the next iteration.
-                            state.NotePartComplete(CompletionPart.MembersCompleted);
-                            break;
+
+                            if (state.NotePartComplete(CompletionPart.MembersCompletedChecksStarted))
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:637]: nit: double empty line above

File: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs (hunk 7 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:1413]: https://github.com/dotnet/roslyn/pull/61436/files#diff-edd6e15a777d20344e841412cb37cad70aa68691075fc9d77fa12b630377e2c9R1262-R1285

**Comment by 333fred** [comment]
> Can we update InvalidModifierCombinations with other invalid combination of modifiers (for example ref readonly and ref) listed in the spec update?

Already covered by `RefReturningProperties`, however I will add a `ref readonly` example to that test.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 8 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:637]: nit: The name of the completion flag is odd, as it suggests it comes in play when members are completed, when in fact it comes before setting `CompletionPart.MembersCompleted`. Consider `MembersChecksStarted` or `MembersCompletionChecksStarted`.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 9]
```diff
@@ -632,11 +633,22 @@ internal override void ForceComplete(SourceLocation? locationOpt, CancellationTo
...
+                            {
+                                cancellationToken.ThrowIfCancellationRequested();
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:639]: This cancellation token check seems dangerous. If we get here and hit a cancellation, it seems we'll never get back into this block because we've already set `CompletionPart.MembersCompletedChecksStarted`.

Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:639]: I've move this before the `if`, you're right that it's dangerous.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 8 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:637]: > we start the checks we perform after members are completed.

There is a rub here. I that the `MembersCompleted` completion part to indicate when the members are completed. The checks we're adding come just before members are completed.

Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:637]: > The checks we're adding come just before members are completed.

The members are all completed when these checks are run. We don't publish that fact yet, because we want to run some checks now that they are completed, but they are completed.

File: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs (hunk 7 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:1413]: Thanks. I see that covers `set` with various accessibility. Do we also have missing setter/initer? (I searched but couldn't find it)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 8 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:637]: Yes, I get that we have the members at hand. But we can't say they are completed until we've marked that part as complete. That's the definition of `CompletionPart.MembersCompleted`.
Perhaps `CompletionPart.StartValidatingMembers`?

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 6)

File: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs (hunk 7 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:1413]: That same test covers that as well.

**Comment by jcouv** [comment]
> Already covered by RefReturningProperties, however I will add a ref readonly example to that test.

The spec says "`required` cannot be combined with the following modifiers: `ref` - when added for ref fields", but that is a test for a ref-returning property. There seems to be a discrepancy.

File: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs (hunk 7 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:1413]: Cool. Thanks!

File: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs [hunk 10]
```diff
@@ -1242,63 +1242,186 @@ public void RefReturningProperties()
 class C
 {
     private int i;
-    public required ref int Prop => ref i;
+    public required ref int Prop1 => ref i;
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:1245]: Based on this, I think the spec should say "`ref` - when added for ref ~~fields~~ properties".

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 7)

File: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs (hunk 10 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:1245]: Yeah, I'll send an update.
